pkg/localstorev2/internal/reserve/reserve.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package reserve
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/localstorev2/internal"
+	"github.com/ethersphere/bee/pkg/log"
+	storage "github.com/ethersphere/bee/pkg/storage"
+	storagev2 "github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology"
+)
+
+// loggerName is the tree path name of the logger for this package.
+const loggerName = "reserve"
+
+const storageRadiusKey = "reserve_storage_radius"
+const DefaultRadiusWakeUpTime = time.Minute * 5
+
+type Reserve struct {
+	mtx sync.Mutex
+
+	baseAddr     swarm.Address
+	stateStore   storage.StateStorer
+	radiusSetter topology.SetStorageRadiuser
+	logger       log.Logger
+
+	capacity int
+	size     int
+	radius   uint8
+}
+
+type Sample struct {
+	Items []swarm.Address
+	Hash  swarm.Address
+}
+
+/*
+	pull by 	bin - binID
+	evict by 	bin - batchID
+	sample by 	bin
+*/
+
+func New(baseAddr swarm.Address, store storagev2.Store, capacity int, reserveRadius uint8, stateStore storage.StateStorer, radiusSetter topology.SetStorageRadiuser, logger log.Logger) (*Reserve, error) {
+
+	rs := &Reserve{
+		baseAddr:     baseAddr,
+		capacity:     capacity,
+		stateStore:   stateStore,
+		radiusSetter: radiusSetter,
+		logger:       logger.WithName(loggerName).Register(),
+	}
+
+	var radius uint8
+	err := stateStore.Get(storageRadiusKey, &radius)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) { // fresh node
+			err = rs.SetRadius(reserveRadius)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	}
+
+	var size int
+	err = store.Iterate(storagev2.Query{
+		Factory: func() storagev2.Item {
+			return &batchRadiusItem{}
+		},
+		Prefix:        string(radius),
+		PrefixAtStart: true,
+	}, func(res storagev2.Result) (bool, error) {
+		size++
+		return false, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	rs.size = size
+
+	return rs, nil
+}
+
+func (r *Reserve) Putter(store internal.Storage) storagev2.Putter {
+
+	return storagev2.PutterFunc(func(ctx context.Context, chunk swarm.Chunk) error {
```

Comment:
For the putter we will need to do 2 more things:
1. First is to check for index collisions. You can refer to the internal/upload package for example.
2. We need to use the new package @mrekucci is adding to store stamps. We have found some issue with having stamps in chunkstore, so the reserve will need to manage its own stamps basically.

Reply:
We need to sync about this.